src/utils.c

```diff
@@ -743,3 +743,169 @@ gboolean r_fakeroot_cleanup(const gchar *envpath, GError **error)
+	 * then parse the remaining 'version core'
+	 */
+
+	build_pos = g_strrstr(version_copy, "+");
+	if (build_pos != NULL) {
+		if (build != NULL) {
+			*build = g_strdup(build_pos + 1);
+			if (*build == NULL)
+				goto out;
+		}
+		*build_pos = '\0';
+	}
+
+	pre_release_pos = g_strrstr(version_copy, "-");
+	if (pre_release_pos != NULL) {
+		if (pre_release != NULL) {
+			*pre_release = g_strdup(pre_release_pos + 1);
+			if (*pre_release == NULL)
+				goto out;
+		}
+		*pre_release_pos = '\0';
+	}
+
+	version_parts = g_strsplit(version_copy, ".", 3);
+	// Note that g_strsplit returns an empty array when splitting: ""
+	if (version_parts == NULL)
+		goto out;
+	// and NULL terminates arrays it creates
+	if (version_parts[0] == NULL)
+		goto out;
+	// only on component -> shift right
+	if (version_parts[1] == NULL) {
+		version_parts[2] = version_parts[0];
+		version_parts[0] = NULL;
+	}
+	// only two components -> shift right
+	if (version_parts[1] != NULL && version_parts[2] == NULL) {
+		version_parts[2] = version_parts[1];
+		version_parts[1] = version_parts[0];
+		version_parts[0] = NULL;
+	}
+
+	for (int i = 0; i < 3; i++) {
+		if (version_parts[i] != NULL) {
+			// Note that negative numbers are already implicitly excluded due to the search for the '-pre_release'
+			version_core[i] = g_ascii_strtoll(version_parts[i], NULL, 10);
+			// handle g_ascii_strtoll returning zero on a conversion error
+			if ((version_core[i] == 0) && (g_strcmp0(version_parts[i], "0") != 0)) {
+				g_strfreev(version_parts);
+				goto out;
+			}
+		} else
+			version_core[i] = 0;
+	}
+
+	g_strfreev(version_parts);
+	return TRUE;
+
+out:
+	if (pre_release && *pre_release)
+		g_free(*pre_release);
+	if (build && *build)
+		g_free(*build);
+	return FALSE;
+}
+
+gboolean semver_less_equal(const guint version_a[3], gchar *pre_release_a, const guint version_b[3], gchar *pre_release_b)
+{
+	gchar **pre_fields_a = NULL;
+	gchar **pre_fields_b = NULL;
+	gboolean is_num_a = FALSE;
+	gboolean is_num_b = FALSE;
+	guint num_a = 0;
+	guint num_b = 0;
+	gint cmp = 0;
+	gboolean result = FALSE;
+	int i = 0;
+
+	// compare version cores: major, minor, patch
+	for (i = 0; i < 3; i++) {
+		if (version_a[i] < version_b[i]) {
+			return TRUE;
+		} else if (version_a[i] > version_b[i]) {
+			return FALSE;
+		}
+	}
+
+	// version cores are equal, compare pre-release identifiers
+	if (pre_release_a == NULL && pre_release_b == NULL) {
+		return TRUE;
+	} else if (pre_release_a == NULL) {
+		return FALSE; // version_a > version_b-pre_release
+	} else if (pre_release_b == NULL) {
+		return TRUE; // version_a-pre_release < version_b
+	}
+
+	// compare	dot-separated fields of pre-release identifiers
+	pre_fields_a = g_strsplit(pre_release_a, ".", 0);
+	pre_fields_b = g_strsplit(pre_release_b, ".", 0);
+
+	i = 0;
+	while (pre_fields_a[i] != NULL && pre_fields_b[i] != NULL) {
+		is_num_a = g_ascii_isdigit(pre_fields_a[i][0]);
+		is_num_b = g_ascii_isdigit(pre_fields_b[i][0]);
```

Comment:
The SemVer spec requires that "Identifiers consisting of only digits are compared numerically", but this only checks the first character. Using `g_ascii_string_to_unsigned` here could solve that, as the different cases a reported via `GErrors` that can be handled explicitly.

Reply:
not quite - this only checks if a number starts here = do we need to compare numerically or lexicographically; later the "whole" number is parsed an compared

Reply:
Hmm, if we have `1foo` in `pre_fields_a[i]` and a valid number in `pre_fields_b[i]`, `is_num_a` and `is_num_b` would be `TRUE`. Then we'd use the "compare numerically" branch, where `g_ascii_strtoull` would ignore the `foo` suffix.

I read the spec as in this case, they are not "consisting of only digits", so they need to be compared in ASCII order.

Reply:
merde! you're absolutely right... regex to the rescue ;-)

Reply:
Why not the approach with `g_ascii_string_to_unsigned` as I suggested above?

Reply:
am/was a bit reluctant since the rauc codebase does not use it (yet...) and that the min/max have to be set, only to end up at 0/MAXUINT; but the error return like you mentioned could be worth it - giving it a shot (-:

Reply:
It's already included in the minimum glib version we required, so I don't see a problem with using it.

Reply:
ACK, switched to string_to_unsigned with v4